chain/src/main/scala/org/bitcoins/chain/blockchain/ChainHandler.scala

```diff
@@ -357,6 +359,90 @@ case class ChainHandler(
   override def getFiltersAtHeight(height: Int)(
       implicit ec: ExecutionContext): Future[Vector[CompactFilterDb]] =
     filterDAO.getAtHeight(height)
+
+  /** Implements [[ChainApi.getMatchingBlocks()]].
+    *
+    * I queries the filter database for [[batchSize]] filters a time
+    * and tries to run [[GolombFilter.matchesAny]] for each filter in a separate thread.
+    *
+    * For best results use a separate execution context.
+    */
+  def getMatchingBlocks(
+      addresses: Vector[BitcoinAddress],
+      startOpt: Option[BlockStamp] = None,
+      endOpt: Option[BlockStamp] = None,
+      batchSize: Int = chainConfig.filterBatchSize,
+      parallelismLevel: Int = Runtime.getRuntime.availableProcessors())(
+      implicit ec: ExecutionContext): Future[Vector[DoubleSha256DigestBE]] = {
+
+    logger.info(
+      s"Starting looking for matching blocks for addresses ${addresses.mkString(",")}")
+
+    val bytes = addresses.map(_.scriptPubKey.asmBytes)
+
+    /** Iterates over the filters in the range to find matches */
+    @tailrec
+    def loop(
+        start: Int,
+        end: Int,
+        acc: Future[Vector[DoubleSha256DigestBE]]): Future[
+      Vector[DoubleSha256DigestBE]] = {
+      if (end <= start) {
+        acc
+      } else {
+        val startHeight = end - (batchSize - 1)
+        val endHeight = end
+        val newAcc: Future[Vector[DoubleSha256DigestBE]] = for {
+          compactFilterDbs <- filterDAO.getBetweenHeights(startHeight,
+                                                          endHeight)
+          groupSize = if (compactFilterDbs.size / parallelismLevel * parallelismLevel < compactFilterDbs.size)
```

Comment:
so i implemented two helper functions here so that we can do this:

```scala
          groupSize = calcGroupSize(compactFilterDbs,parallelismLevel)
          grouped = compactFilterDbs.grouped(groupSize)
          filtered <- findMatches(grouped,scriptBytes)
```

and here are the helper functions:

```scala
  //scaladoc...
  private def calcGroupSize(compactFilterDbs: Vector[CompactFilterDb], parallelismLevel: Int): Int = {
    if (compactFilterDbs.size / parallelismLevel * parallelismLevel < compactFilterDbs.size)
      compactFilterDbs.size / parallelismLevel + 1
    else compactFilterDbs.size / parallelismLevel
  }

  //scaladoc...
  private def findMatches(grouped: Iterator[Vector[CompactFilterDb]],scriptBytes: Vector[ByteVector])(implicit ec: ExecutionContext): Future[Vector[DoubleSha256DigestBE]] = {
    val allMatchesF: Future[Iterator[Vector[DoubleSha256DigestBE]]] = {
      val nestedMatches: Iterator[Future[Vector[DoubleSha256DigestBE]]] = grouped.map { filterGroup =>
        //matchings for the subgroup
        val nestedMatch: Future[Vector[DoubleSha256DigestBE]] = {
          //we need to wrap in a future here to make sure we can
          //run these matches in parallel
          Future {
            filterGroup.foldLeft(Vector.empty[DoubleSha256DigestBE]) {
              (blocks, filter) =>
                if (filter.golombFilter.matchesAny(scriptBytes))
                  blocks :+ filter.blockHashBE
                else blocks
            }
          }
        }

        nestedMatch
      }
      Future.sequence(nestedMatches)
    }
    allMatchesF.map(_.flatten.toVector)
  }
```

feel free to remove type annotations inside of the helper methods if you want as it does clutter things up.